### PR TITLE
EWB-2446: Add example traces

### PR DESCRIPTION
**Pull request by Lax125**
# Description

Add examples of using tracing, specifically the traversal functions in the `tracing` module.

# Associated tasks:

Tasks blocking this one:
- [x] https://github.com/zepben/pp-translator/pull/19
- [x] https://github.com/zepben/edith-sdk/pull/3
- [x] https://github.com/zepben/eas-client-python/pull/10

Tasks this is blocking:
- None

# Checklist:

- [x] I have performed a self review of my own code.
- [x] I have added/updated unit tests for these changes, and if not I have explained why they are not necessary.
- [x] I have updated the changelog.
- [x] I have handled all new warnings generated by the compiler or IDE.

### CODE REVIEW

**Comment by kgreav** [comment]
Sorry merged your other PR and will probably cause you some rebase pain here...
You'll want to use `git rebase -i --onto main <branch>` or something similar to make sure you only get the commits from this branch.

File: src/zepben/examples/fetching_network_model.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+#  License, v. 2.0. If a copy of the MPL was not distributed with this
+#  file, You can obtain one at https://mozilla.org/MPL/2.0/.
+from zepben.evolve import Conductor, PowerTransformer, connect_with_password, SyncNetworkConsumerClient, ConductingEquipment, EnergyConsumer, Switch
+from zepben.protobuf.nc.nc_requests_pb2 import INCLUDE_ENERGIZED_LV_FEEDERS, INCLUDE_ENERGIZED_FEEDERS, INCLUDE_ENERGIZING_SUBSTATIONS, \
+    INCLUDE_ENERGIZING_FEEDERS
+
+
+def main():
+    # See connecting_to_grpc_service.py for examples of each connect function
+    channel = connect_with_password(host="EWB hostname",
```
[the remark this answers is not in the document]
Comment (Lax125) [reply src/zepben/examples/fetching_network_model.py]: No actual password here.

[the remark this answers is not in the document]
Comment (Lax125) [reply src/zepben/examples/fetching_network_model.py]: @sonatype-lift ignore

**Comment by Lax125** [comment]
No tests needed for example code.

File: src/zepben/examples/tracing.py [hunk 2]
```diff
@@ -0,0 +1,381 @@
...
+    visited.clear()
+
+    # There is also a breadth-first version, which guarantees that each equipment is visited at most once.
+    print("Connected Equipment Breadth Trace:")
+    await connected_equipment_breadth_trace().add_step_action(print_step).run(start_item)
+    print(f"Number of equipment visited: {len(visited)}")
+    print()
+    visited.clear()
+
+    # The normal connected equipment trace iterates through all equipment normally connected to the starting equipment.
```
Comment (kgreav) [inline src/zepben/examples/tracing.py:71]: reword this to "all equipment connected to the starting equipment in the networks' normal state"

File: src/zepben/examples/tracing.py [hunk 3]
```diff
@@ -0,0 +1,381 @@
...
+    # By setting the switch from node 671 to 692 to normally open on at least one phase, the traversal will not trace through the switch.
+    network.get("sw_671_692", Switch).set_normally_open(True, phase=SinglePhaseKind.A)
```
Comment (kgreav) [inline src/zepben/examples/tracing.py:73]: so even if other phases are closed on the switch this trace doesn't trace through them? if that's the case can you mention that explicitly?
seems like a weird behaviour though...

File: src/zepben/examples/tracing.py [hunk 4]
```diff
@@ -0,0 +1,381 @@
...
+    print()
+    print("Normal Connected Equipment Trace:")
+    await normal_connected_equipment_trace().add_step_action(print_step).run(start_item)
+    print(f"Number of equipment visited: {len(visited)}")
+    print()
+    visited.clear()
+
+    # The normal connected equipment trace iterates through all equipment normally connected to the starting equipment.
+    # By setting the switch from node 671 to 692 to currently open on at least one phase, the traversal will not trace through the switch.
+    switch.set_normally_open(True, phase=SinglePhaseKind.B)
```
Comment (kgreav) [inline src/zepben/examples/tracing.py:84]: looks like you've typo'd normal/normally here and it should be current/currently.

Probably worth rewording the comment like i've suggested above as well.

File: src/zepben/examples/tracing.py [hunk 5]
```diff
@@ -0,0 +1,381 @@
...
+
+    # A breadth-first connectivity trace is also available.
+    print("Connectivity Breadth Trace:")
+    await connectivity_breadth_trace().add_step_action(print_connectivity).run(start_item)
+    print(f"Number of connectivities visited: {len(visited)}")
+    print()
+    visited.clear()
+
+    # The normal connectivity trace is analogous to the normal connected equipment trace,
+    # and likewise does not go through switches with at least open phase.
```
Comment (kgreav) [inline src/zepben/examples/tracing.py:124]: typo: at least one open phase

**Review by kgreav — CHANGES_REQUESTED** [review]
very good sir - only minor comments/questions.

we can probably merge it after you make the fixes and then get krishneel to test/document them based off this in the docusaurus repo, and any fixes/improvements he makes we can incorporate here.

File: src/zepben/examples/tracing.py (hunk 5 above)
Comment (Lax125) [reply src/zepben/examples/tracing.py]: Fixed

File: src/zepben/examples/tracing.py (hunk 3 above)
Comment (Lax125) [reply src/zepben/examples/tracing.py:74]: It's how tracing works on the JVM sdk version too AFAIK, so this doesn't seem like a bug. I think phase traces are used instead in cases where you do want it to go through.

File: src/zepben/examples/tracing.py (hunk 4 above)
Comment (Lax125) [reply src/zepben/examples/tracing.py]: Fixed typo and reworded.

File: src/zepben/examples/tracing.py (hunk 2 above)
Comment (Lax125) [reply src/zepben/examples/tracing.py]: Done.

**Review by kgreav — APPROVED** [review]
